tap_bing_ads/__init__.py

```diff
@@ -644,20 +644,63 @@ def stream_report(stream_name, report_name, url, report_time):
 
 async def sync_report(client, account_id, report_stream):
     report_name = stringcase.pascalcase(report_stream.stream)
-
-    report_schema = get_report_schema(client, report_name)
-    singer.write_schema(report_stream.stream, report_schema, [])
-
     state_key = '{}_{}'.format(account_id, report_stream.stream)
-    config_start_date = CONFIG.get('start_date')
+
     bookmark = singer.get_bookmark(STATE,
                                    state_key,
                                    'date')
+    config_start_date = CONFIG.get('start_date')
     conversion_window = int(CONFIG.get('conversion_window', '-30'))
-    start_date = arrow.get(bookmark or config_start_date).shift(days=conversion_window)
-    end_date = arrow.get(CONFIG.get('end_date')) # defaults to now
+    start_date = arrow.get(bookmark or config_start_date) \
+                      .shift(days=conversion_window)
+    end_date = arrow.get(CONFIG.get('end_date'))  # defaults to now
+
+    report_schema = get_report_schema(client, report_name)
+    singer.write_schema(report_stream.stream, report_schema, [])
+
+    report_time = arrow.get().isoformat()
+
+    request_id = get_report_request_id(client, account_id, report_stream,
+                                       report_name, start_date, end_date,
+                                       state_key)
+
+    singer.write_bookmark(STATE, state_key, 'request_id', request_id)
```

Comment:
Should there be a call to `singer.write_state()` after this to persist the request_id in case of failure?

Reply:
Yes - writing it as soon as possible is ideal.

Reply:
Updated in the latest commit.